agent/consul/stream/event_snapshot.go

```diff
@@ -49,34 +49,33 @@ func (s *eventSnapshot) appendAndSplice(req SubscribeRequest, fn SnapshotFunc, t
 func (s *eventSnapshot) spliceFromTopicBuffer(topicBufferHead *bufferItem, idx uint64) {
 	item := topicBufferHead
 	for {
-		next := item.NextNoBlock()
```

Comment:
If I understand correctly the location of this call is the source of the bug right?

So before with it here there is a case where `topicBufferHead` pointed to an actual event that should be delivered after the snapshot, but this first iteration skips over it on to the next one and so misses that event?

I guess the assumption here is that because we took a reference to `topicBufferHead` _before_ we made the snapshot, the item it points to is either a nil link-only item (because it was empty), or will be an event that must already be in the snapshot.

I'm interested to know under which circumstances that invariant can be broken and the `topicBufferHead` actually point to an event that is _after_ the snapshot index in your tests.

That said, the changes here all seem reasonable and there's no reason not to have this more robust case _even_ if the original invariant assumptions did hold.

Reply:
Yup, that all sounds right. The test that triggered this condition returned an `index=1` which was older than the topic buffer head.  I'm looking around now to see if we document this invariant. I imagine it will be important for future topics to understanding what index they need to send back from the snapshot functions. I think you are right that this isn't a bug that can impact users today because the current snapshot functions will always return the index of the topic buffer head.